numba/runtime/nrt.c

```diff
@@ -0,0 +1,271 @@
+#include <stdarg.h>
+#include <string.h> /* for memset */
+#include "nrt.h"
+#include "assert.h"
+
+union MemInfo{
+    struct {
+        size_t         refct;
+        dtor_function  dtor;
+        void          *dtor_info;
+        void          *data;
+        size_t         size;    /* only used for NRT allocated memory */
+    } payload;
+
+    /* Freelist or Deferred-dtor-list */
+    MemInfo *list_next;
+};
+
+
+struct MemSys{
+    /* Ununsed MemInfo are recycled here */
+    MemInfo * volatile mi_freelist;
+    /* MemInfo with deferred dtor */
+    MemInfo * volatile mi_deferlist;
+    /* Atomic increment and decrement function */
+    atomic_inc_dec_func atomic_inc, atomic_dec;
+    /* Atomic CAS */
+    atomic_cas_func atomic_cas;
+    /* Shutdown flag */
+    int shutting;
+
+};
+
+/* The Memory System object */
+static MemSys TheMSys;
+
+
+static
+MemInfo *nrt_pop_meminfo_list(MemInfo * volatile *list) {
+    MemInfo *old, *repl, *head;
+
+    head = *list;     /* get the current head */
```

Comment:
You certainly want this inside the loop (since another thread may change the head pointer).